### PR TITLE
AST string representation

### PR DESCRIPTION
**Pull request by VladislavsPerkanuks**
AST to MF2 string representation

~**Waiting for #5**~

### CODE REVIEW

File: ast.go [hunk 1]
```diff
@@ -1,7 +1,33 @@
...
+		Message: SimpleMessage{
+			Patterns: []Pattern{
+				TextPattern("Hello, "),
+				PlaceholderPattern{Expression: VariableExpression{Variable: Variable("variable")}},
+				TextPattern(" World!"),
+			},
+		},
+	}
+
+	fmt.Print(ast) // Hello, { $variable } World!
```
[the remark this answers is not in the document]
Comment (VladislavsPerkanuks) [reply ast.go:28]:
```abnf
variable-expression = "{" [s] variable [s annotation] [s] "}"
; where s is whitespace and [] is optional, so [s] == optional whitespace
```
so both options are correct
```
Hello, { $variable } World!
Hello, {$variable} World!
```

File: ast.go [hunk 2]
```diff
@@ -111,43 +161,74 @@ type LiteralExpression struct {
...
 
 type NumberLiteral float64
 
-func (NumberLiteral) node()     {}
-func (NumberLiteral) literal()  {}
-func (NumberLiteral) unquoted() {}
+func (NumberLiteral) node()             {}
+func (NumberLiteral) literal()          {}
+func (NumberLiteral) unquoted()         {}
+func (nl NumberLiteral) String() string { return fmt.Sprint(float64(nl)) }
```
Comment (klavinsjanis) [inline ast.go:231]: is type conversion to float64 necessary?

Comment (VladislavsPerkanuks) [reply ast.go:231]:
```go
fmt.Sprint(nl)
```

This one will go into infinite recursion :)

File: parse_test.go [hunk 3]
```diff
@@ -327,7 +328,22 @@ func TestParseSimpleMessage(t *testing.T) {
 			actual, err := Parse(tt.input)
 			require.NoError(t, err)
 
-			require.Equal(t, tt.expected, actual)
+			// Check that AST message is equal to expected one.
+			require.Equal(t, tt.expected, actual.Message)
+
+			// Check that AST message converted back to string is equal to input.
+
+			// Edge case: scientific notation number is converted to normal notation, hence comparison is bound to fail.
```
Comment (klavinsjanis) [inline parse_test.go:336]: I wonder if there is any way to preserve the notation

Comment (VladislavsPerkanuks) [reply parse_test.go:336]: in token we store string representation of numberLIteral
in node we store float representation of numberLiteral, using json.Unmarshal

so json.Unmarshal converts "1e3" to 1000.0

going back to string we don't have any information about the original notation.

So simple answer: it is not possible

Comment (klavinsjanis) [reply parse_test.go:336]: Maybe we could store it as a string in node?

File: ast.go [hunk 4]
```diff
@@ -248,16 +369,52 @@ type Identifier struct {
...
+
+// sliceToString converts a slice of Nodes to a string, separated by sep.
+func sliceToString[T Node](s []T, sep string) string {
+	nodeStrings := make([]string, len(s))
+	for i, node := range s {
+		nodeStrings[i] = fmt.Sprint(node)
+	}
+
+	return strings.Join(nodeStrings, sep)
+}
```
Comment (janishorsts) [inline ast.go:420]: more efficient

```suggestion
func sliceToString[T Node](s []T, sep string) string {
	if len(s) == 0 {
		return ""
	}
	
	r := fmt.Sprint(s[0])
	
	for _, v := range s[1:] {
		r += sep + fmt.Sprint(v)
	}
	
	return r
}
```

File: ast.go [hunk 5]
```diff
@@ -158,18 +239,30 @@ type FunctionAnnotation struct {
...
+	return fmt.Sprintf("%s %s", fa.Function, sliceToString(fa.Options, " "))
+}
+
 type PrivateUseAnnotation struct {
 	Annotation
 
 	// TODO: Implementation
 }
 
+func (PrivateUseAnnotation) String() string { return "TODO" }
```
Comment (janishorsts) [inline ast.go:256]: This breaks MF2 syntax. Use something like `.|private annotation not implemented|`.

**Review by janishorsts — APPROVED** [review]

**Review by DarjaLeonova — APPROVED** [review]

**Review by klavinsjanis — APPROVED** [review]

**Pull request merged**